nbextensions/bulkImportCell/main.js

```diff
@@ -0,0 +1,113 @@
+/*
+ * KBase Bulk Import Cell Extension
+ *
+ * Supports kbase bulk import cells
+ *
+ * Note that, out of our control, this operates under the "module as singleton" model.
+ * In this model, the execution of the module the first time per session is the same
+ * as creating a global management object.
+ *
+ * @param {type} $
+ * @param {type} Jupyter
+ * @param {type} html
+ *
+ */
+define([
+    'jquery',
+    'base/js/namespace',
+    'bluebird',
+    'common/error',
+    './bulkImportCell',
+    'custom/custom'
+], function(
+    $,
+    Jupyter,
+    Promise,
+    Error,
+    BulkImportCell
+) {
+    'use strict';
+    const CELL_TYPE = 'app-bulk-import';
+
+    /**
+     * This gets called on extension initialization. This iterates over all existing cells,
+     * and if it's a bulk import cell, then we init the wrapping BulkImportCell class.
+     */
+    function setupNotebook() {
+        return Promise.all(Jupyter.notebook.get_cells().map((cell) => {
+            if (BulkImportCell.isBulkImportCell(cell)) {
+                try {
+                    new BulkImportCell(cell);
+                }
+                catch(error) {
+                    // If we have an error here, there is a serious problem setting up the cell and it is not usable.
+                    // What to do? The safest thing to do is inform the user, and then strip out the cell, leaving
+                    // in it's place a markdown cell with the error info.
+                    // For now, just pop up an error dialog;
+
+                    Error.reportCellError('Error starting bulk import cell',
+                        'There was an error starting the bulk import cell',
+                        error);
+                }
+            }
+        }));
+    }
+
+    /*
+     * Called directly by Jupyter during the notebook startup process.
+     * Called after the notebook is loaded and the dom structure is created.
+     * The job of this call is to mutate the notebook and cells to be aware of
+     * the bulk import cells.
+     */
+    function load_ipython_extension() {
+        setupNotebook()
+            .then(() => {
+                $([Jupyter.events]).on('insertedAtIndex.Cell', (event, payload) => {
+                    const cell = payload.cell,
+                        setupData = payload.data,
+                        jupyterCellType = payload.type;
+
+                    if (jupyterCellType !== 'code' ||
+                        !setupData ||
+                        !(setupData.type === CELL_TYPE)) {
+                        return;
```

Comment:
should this return an error?

Reply:
Nope. That's the nature of the pattern we're using in these extensions. Each cell extension is independent and cycles over the notebook cells at startup. If it finds a cell it's not supposed to turn into, say, a Bulk Import Cell, or an App Cell, it just carries on.

Reply:
As we accumulate more of these, it might make sense to have one heavier controller extension that acts as an overall cell specializer, so they only get iterated over once. But it's a quick process and hasn't been a problem yet. And that's a little out of scope for this PR.